src/main/java/de/tum/in/www1/artemis/service/notifications/push_notifications/PushNotificationService.java

```diff
@@ -0,0 +1,130 @@
+    }
+
+    private static final Logger log = LoggerFactory.getLogger(PushNotificationService.class);
+
+    private static final Gson gson = new Gson();
+
+    /**
+     * Send all the notifications requests to the endpoint. Potentially, optimize the sending using a batched request.
+     *
+     * @param requests     the requests previously built using buildSendRequest
+     * @param relayBaseUrl the url of the relay
+     */
+    abstract void sendNotificationRequestsToEndpoint(List<RelayNotificationRequest> requests, String relayBaseUrl);
+
+    @Override
+    public final void sendNotification(Notification notification, User user, Object notificationSubject) {
+        sendNotification(notification, Collections.singletonList(user), notificationSubject);
+    }
+
+    @Override
+    @Async
+    public void sendNotification(Notification notification, List<User> users, Object notificationSubject) {
+        final Optional<String> relayServerBaseUrl = getRelayBaseUrl();
+
+        if (relayServerBaseUrl.isEmpty())
+            return;
+
+        final NotificationType type = NotificationConstants.findCorrespondingNotificationType(notification.getTitle());
+
+        final List<PushNotificationDeviceConfiguration> userDeviceConfigurations = getRepository().findByUserIn(users, getDeviceType());
+        if (userDeviceConfigurations.isEmpty())
+            return;
+
+        final String date = Instant.now().toString();
+        final String payload = gson.toJson(new PushNotificationData(notification.getTransientPlaceholderValuesAsArray(), notification.getTarget(), type.name(), date));
+
+        final byte[] iv = new byte[16];
+
+        List<RelayNotificationRequest> notificationRequests = userDeviceConfigurations.stream().flatMap(deviceConfiguration -> {
+            random.nextBytes(iv);
+
+            SecretKey key = new SecretKeySpec(deviceConfiguration.getSecretKey(), "AES");
+
+            String ivAsString = Base64.getEncoder().encodeToString(iv);
+            Optional<String> payloadCiphertext = encrypt(payload, key, iv);
+
+            return payloadCiphertext.stream().map(s -> new RelayNotificationRequest(ivAsString, s, deviceConfiguration.getToken()));
+        }).toList();
+
+        sendNotificationRequestsToEndpoint(notificationRequests, relayServerBaseUrl.get());
+    }
+
+    protected abstract PushNotificationDeviceConfigurationRepository getRepository();
+
+    abstract PushNotificationDeviceType getDeviceType();
+
+    abstract Optional<String> getRelayBaseUrl();
+
+    record PushNotificationData(String[] notificationPlaceholders, String target, String type, String date) {
+    }
+
+    /**
+     * Perform symmetric AES encryption.
+     *
+     * @param payload the text to encrypt
+     * @param key     the secret key to encrypt with
+     * @param iv      the initialization vector needed for CBC
+     * @return the ciphertext
+     */
+    private static Optional<String> encrypt(String payload, SecretKey key, byte[] iv) {
```

Comment:
```suggestion
    private static Optional<String> encrypt(String payload, SecretKey key, byte[] initVector) {
```
Aha, the Javadoc explains it.